File: day_3/puzzle_2/solve.rb
# frozen_string_literal: true

require_relative '../../benchmark/benchmark'

Benchmark.run('day_3_puzzle_2') do
  def letter_to_priority(letter)
    char_code = letter.ord
    char_code >= 97 ? char_code - 96 : char_code - 38
  end

  File.open('./day_3/input.txt', 'r') do |file|
    ruck_sack_groups = file.read.split("\n").each_with_index.each_with_object([]) do |(ruck_sack, i), groups|
      groups.push([]) if (i % 3).zero?
      groups.last.push(ruck_sack)
    end

    priority = ruck_sack_groups.inject(0) do |sum, group|
      group.first.each_char do |char|
        if group[1].include?(char) && group[2].include?(char)
          sum += letter_to_priority(char)
          break
        end
      end

      sum
    end

    priority
  end
end

File: day_3/puzzle_2/solve.ts
import fs from "fs";
import { benchmark } from "../../benchmark/benchmark";

benchmark(() => {
  function letterToPriority(letter: string) {
    const charCode = letter.charCodeAt(0);
    return charCode >= 97 ? charCode - 96 : charCode - 38;
  }

  const file = fs.readFileSync("./day_3/input.txt", { encoding: "utf-8" });

  const ruckSackGroups = file.split("\n").reduce((groups, ruckSack, i) => {
    if (i % 3 === 0) {
      groups.push([]);
    }

    groups.at(-1)?.push(ruckSack);

    return groups;
  }, [] as string[][]);

  const priority = ruckSackGroups.reduce((sum, group) => {
    for (const char of group[0]) {
      if (group[1].includes(char) && group[2].includes(char)) {
        sum += letterToPriority(char);
        break;
      }
    }

    return sum;
  }, 0);

  return priority;
}, "day_3_puzzle_2");
